src/OrleansRuntime/Core/Dispatcher.cs

```diff
@@ -625,23 +625,61 @@ private bool ShouldLogError(Exception ex)
                    !(ex.GetBaseException() is ClientNotAvailableException);
         }
 
-        // this is a compatibility method for portions of the code base that don't use
-        // async/await yet, which is almost everything. there's no liability to discarding the
-        // Task returned by AsyncSendMessage()
-        internal void SendMessage(Message message, ActivationData sendingActivation = null)
-        {
-            AsyncSendMessage(message, sendingActivation).Ignore();
-        }
-
-        /// <summary>
-        /// Resolve target address for a message
-        /// - use transaction info
-        /// - check ordering info in message and sending activation
-        /// - use sender's placement strategy
-        /// </summary>
-        /// <param name="message"></param>
-        /// <returns>Resolve when message is addressed (modifies message fields)</returns>
-        private async Task AddressMessage(Message message)
+		// this is a compatibility method for portions of the code base that don't use
+		// async/await yet, which is almost everything. there's no liability to discarding the
+		// Task returned by AsyncSendMessage()
+		internal void SendMessage(Message message, ActivationData sendingActivation = null)
+		{
+			if (TrySendMessageToLocalActivation(message))
```

Comment:
Yes, this is most questionable additional optimization and your question was my primary concern when doing it, but it does not relates to original PR intention. 
I also saw the comment in  [Catalog's FastLookup method](https://github.com/dVakulen/orleans/blob/master/src/OrleansRuntime/Catalog/Catalog.cs#L1217). This method tries to get information from local ```ActivationDirectory``` cache and is being used by ```RandomPlacementDirector``` , ```StatelessWorkerDirector``` and ```ClientObserversPlacementDirector``` on activation selection path; that's why I decided to use it directly instead of going through multiple awaits in PlacementDirectors logic.